scipy/optimize/lbfgsb.py

```diff
@@ -393,10 +434,26 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
     hess_inv = LbfgsInvHessProduct(s[:n_corrs], y[:n_corrs])
 
     task_str = task_str.decode()
-    return OptimizeResult(fun=f, jac=g, nfev=sf.nfev,
+
+    # re-expand x and gradient, there might have been equal_bounds
+    full_x = np.copy(full_x0)
+    full_x[~filtered_params] = x
+    grad = np.empty_like(full_x)
+    grad[filtered_params] = np.nan
+    grad[~filtered_params] = g
+    if filtered_params.any():
+        # expand hess_inv. Do this by setting increments in s and y for fixed
+        # values to 0.
+        s_amended = np.zeros((n_corrs, np.size(full_x0)))
+        y_amended = np.zeros_like(s_amended)
+        s_amended[:, ~filtered_params] = hess_inv.sk
+        y_amended[:, ~filtered_params] = hess_inv.yk
+        hess_inv = LbfgsInvHessProduct(s_amended, y_amended)
```

Comment:
I'll need to take a closer look at this later. In the meantime - do you need to create a `LbfgsInvHessProduct` above and use it to create this one, or could you somehow create only this one? Can/should entries be `nan` instead of 0?

Reply:
If these modifications aren't done then `hess_inv.todense()` results in an array that has shape `(np.count_nonzero(filtered_params), np.count_nonzero(filtered_params))`, i.e. the size of the reduced problem. This might be confusing to people that expect it to have a size corresponding to `full_x0`.

I tried to set `s_amended` to 0 for filtered parameters, and `y_amended` to `np.nan`. However, when one calculates `hess_inv.todense()` all the entries come out as `np.nan`, which is not correct either.

So, I'm not sure which behaviour is the worst, or if there is another way of achieving this?

Reply:
> If these modifications aren't done

I know you have to modify it. I just mean can you modify it by changing the inputs to the original call to the initializer (`LbfgsInvHessProduct`) rather than calling the initializer once and using attributes of the object to call it again.

Then again, I'm wondering whether we can salvage this at all. This is the _inverse_ of the Hessian, not the Hessian itself. Suppose we were trying to comput this directly (rather than approximate it as is done in LBFGSB, right?) When we don't take the second partials with respect to the fixed variables,  we would have to `nan` out some rows and columns corresponding to the fixed variables. What would happen when we took the inverse?

Reply:
It's possible to create the inverse hessian with different inputs (i.e. with nan for grad entries, and 0 for sk entries). I don't know how to judge if this creates a valid LbfgsInvHessProduct or not.